sync/client_watch.go

```diff
@@ -0,0 +1,81 @@
+package sync
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-redis/redis/v7"
+	"github.com/testground/sdk-go/runtime"
+	"go.uber.org/zap"
+)
+
+// WatchClient is used by the Testground daemon to monitor all emitted events by the testplans,
+// in particular the terminal events, such as SuccessEvent, FailureEvent and CrashEvent.
+type WatchClient struct {
+	ctx     context.Context
+	rclient *redis.Client
+	log     *zap.SugaredLogger
+}
+
+func NewWatchClient(ctx context.Context, log *zap.SugaredLogger) (*WatchClient, error) {
+	rclient, err := redisClient(ctx, log)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create redis client: %w", err)
+	}
+
+	return &WatchClient{
+		ctx,
+		rclient,
+		log,
+	}, nil
+}
+
+func (w *WatchClient) FetchAllEvents(rp *runtime.RunParams) ([]*runtime.Event, error) {
```

Comment:
The method implicitly assumes that this will be called at the end. If that's the case, we should document it (as it's part of the contract), and instead of doing this undeterministic pagination, we should do an XLen and consume that many elements (paginating, if necessary), to make this method deterministic.

---

Why not move this to a channel-based subscription? We are going to need real-time consumption pretty soon. And we are already doing a funny dance here to paginate, anyway. We don't need to implement indefinite Block and the unblock flow (like the sync service topic subscription does).

Just passing in a context, keeping the block time of 1 second, and looping until the context is closed is enough.

We can also pass a `closeOnTerminalEvt bool` flag that closes the channel when we find a terminal event (crash, success, failure). We can add `Event#IsTerminal() bool` to abstract over that.

That would allow the daemon to do:

```go
wcl := NewWatchClient()
for evt := range wcl.FetchEvents(ctx) {
}
```

Reply:
Right - this method is not guaranteed to exit, if the producer keeps on spamming events.

We should either refactor and just use the `Subscribe` method, or use `XLen` as you suggest.